presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java

```diff
@@ -0,0 +1,667 @@
+            Map<List<String>, RichColumnDescriptor> descriptorsByPath = getDescriptors(fileSchema, requestedSchema);
+            TupleDomain<ColumnDescriptor> parquetTupleDomain = getParquetTupleDomain(descriptorsByPath, effectivePredicate);
+            Predicate parquetPredicate = buildPredicate(requestedSchema, parquetTupleDomain, descriptorsByPath);
+
+            List<BlockMetaData> blocks = new ArrayList<>();
+            for (BlockMetaData block : parquetMetadata.getBlocks()) {
+                long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
+                if ((firstDataPage >= start) && (firstDataPage < (start + length)) &&
+                        predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain, failOnCorruptedParquetStatistics)) {
+                    blocks.add(block);
+                }
+            }
+
+            MessageColumnIO messageColumnIO = getColumnIO(fileSchema, requestedSchema);
+            ParquetReader parquetReader = new ParquetReader(
+                    messageColumnIO,
+                    blocks,
+                    dataSource,
+                    systemMemoryContext,
+                    maxReadBlockSize,
+                    batchReaderEnabled,
+                    verificationEnabled);
+
+            ImmutableList.Builder<String> namesBuilder = ImmutableList.builder();
+            ImmutableList.Builder<Type> prestoTypes = ImmutableList.builder();
+            ImmutableList.Builder<Optional<Field>> internalFields = ImmutableList.builder();
+            for (int columnIndex = 0; columnIndex < regularColumns.size(); columnIndex++) {
+                IcebergColumnHandle column = regularColumns.get(columnIndex);
+                namesBuilder.add(column.getName());
+                org.apache.parquet.schema.Type parquetField = parquetFields.get(columnIndex);
+
+                Type prestoType = column.getType();
+
+                prestoTypes.add(prestoType);
+
+                if (parquetField == null) {
+                    internalFields.add(Optional.empty());
+                }
+                else {
+                    internalFields.add(constructField(column.getType(), messageColumnIO.getChild(parquetField.getName())));
+                }
+            }
+
+            return new ParquetPageSource(parquetReader, prestoTypes.build(), internalFields.build(), namesBuilder.build(), new RuntimeStats());
+        }
+        catch (Exception e) {
+            try {
+                if (dataSource != null) {
+                    dataSource.close();
+                }
+            }
+            catch (IOException ignored) {
+            }
+            if (e instanceof PrestoException) {
+                throw (PrestoException) e;
+            }
+            String message = format("Error opening Iceberg split %s (offset=%s, length=%s): %s", path, start, length, e.getMessage());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format long: start (declared at line 288), length (declared at line 289). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 389. Consider formatting this data with &#39;%d&#39; instead.